### ISSUE TITLE
False positive `:redundant-do` with reader conditionals

### ISSUE THREAD

**Issue by thiru** [clj-kondo/clj-kondo#1722]
[ To keep development of this project going, consider sponsoring. If you are
already a sponsor, thank you! ]

**version**

clj-kondo v2022.04.25

**platform**

native

**problem**

clj-kondo flags a `:redundant-do` warning when one is required.

**repro**

```clojure
(let [_ nil]
  #?(:cljs
     (do
       (println "cljs 1")
       (println "cljs 2"))
     :clj (println "clj 1")))
```

**expected behavior**

There should not be a `:redundant-do` warning in code above.  Without the do there is a compilation error.

**Comment by borkdude** [comment]
I agree that this is annoying enough to fix :)

**Comment by borkdude** [comment]
We should do it for `:redundant-let` as well

**Comment by borkdude** [comment]
Possibly for `:unused-binding` as well. Also, perhaps make configurable to still get warnings if either language branch has warnings.

**Comment by thiru** [comment]
This would be great, thank you!

Our team did come across the `:unused-binding` situation as well.  We have code where only one language branch makes use of some bindings.  To indicate a warning in this case (which is what kondo currently does) seems like a stricter form of linting, whereas not indicating a warning if _at least_ one language branch makes use of the binding seems more lenient and maybe more desirable in some situations?  Is this what you meant around the configurability?

**Comment by borkdude** [comment]
Exactly.

**Comment by borkdude** [comment]
The mechanism is exactly the same. Clj-kondo lints your sources once for all clj branches and then for all cljs branches. It doesn't know that a redundant do in one branch isn't redundant in the other. This is the same for all linters.

**Comment by thiru** [comment]
Awesome, thanks for fixing this so quickly!

**Comment by sumbach** [comment]
Yes, thank you for fixing this @borkdude ! 😄 

A similar situation can also arise with `:unused-referred-var` and `:unused-namespace` where some namespaces or referred vars are used in clj _or_ cljs, but not both. I suspect this is comparable to the `:unused-binding` case -- at first I was annoyed that I had to add more reader conditionals to fix these linter issues. But these issues _could_ indicate legitimate bugs and may not be good to suppress by default.

I'm hoping to find a better way to write the code to avoid the linter issues _and_ the extra reader conditionals. I'm guessing I would not have written the code this way if I had been using clj-kondo at the time -- the root cause of the problem is my code structure, not clj-kondo behavior.

**Comment by sumbach** [comment]
I noticed a couple of related/knock-on issues related to this one. Let me know if you'd rather I report these as separate issues or feature requests.

```
(let [_ nil]
  #?(:cljs
     (do
       (println "cljs 1")
       (do) ;; TODO: should be redundant-do
       (println "cljs 2"))

     :clj
     (println "clj 1")))

(let [_ nil]
  #?(:cljs
     (let [_ nil] ;; TODO: should *NOT* be redundant-let
       (do
         (println "cljs 1")
         (println "cljs 2")))

     :clj
     (println "clj 1")))
```

I suspect the first issue is caused by the redundant-do check being disabled for the entire contents of any reader conditional block; I think the desired behavior is to disable/ignore redundant-do when it appears as the immediate child of a reader conditional.

The second issue probably points toward applying the same exception to redundant-let and redundant-do.
